src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -242,123 +241,97 @@ private unsafe void ReadDirectoryChangesCallback(uint errorCode, uint numBytes,
             }
         }
 
-        private void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
         {
             Debug.Assert(buffer != null);
             Debug.Assert(buffer.Length > 0);
 
-            // Parse each event from the buffer and notify appropriate delegates
-
-            /******
-                Format for the buffer is the following C struct:
-
-                typedef struct _FILE_NOTIFY_INFORMATION {
-                   DWORD NextEntryOffset;
-                   DWORD Action;
-                   DWORD FileNameLength;
-                   WCHAR FileName[1];
-                } FILE_NOTIFY_INFORMATION;
+            ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
-                NOTE1: FileNameLength is length in bytes.
-                NOTE2: The Filename is a Unicode string that's NOT NULL terminated.
-                NOTE3: A NextEntryOffset of zero means that it's the last entry
-            *******/
-
-            // Parse the file notify buffer:
-            int offset = 0;
-            int nextOffset, action, nameLength;
-            string oldName = null;
-            string name = null;
+            // Note that the buffer is pinned for the lifetime of the watcher.
+            // We have to pin again here to get the pointer.
+            Interop.Kernel32.FILE_NOTIFY_INFORMATION* info;
```

Comment:
Could we do all of this with spans and avoid the unsafe code? Something like:
```
FILE_NOTIFY_INFORMATION fileInfo = buffer.Slice(0, Unsafe.SizeOf<FILE_NOTIFY_INFORMATION>()).Cast<FILE_NOTIFY_INFORMATION>()[0];
Span<char> FileName = buffer.Slice(Unsafe.SizeOf<FILE_NOTIFY_INFORMATION>());
```

Reply:
We need the type of change and offset to the next record as well. The filename is an anysize array unfortunately. Doing it this way is simpler than it used to be.